### ISSUE TITLE
Behaviour of complex permittivity and permeability under different dtype in SimulationConfig

### ISSUE THREAD

**Issue by galcerte** [ymahlau/fdtdx#68]
I noticed that, when choosing either `jnp.float32` or `jnp.complex64` as the `dtype` argument in `SimulationConfig` and using complex permittivity (though I figure this happens with permeability too), I get warnings from JAX where there's casting from complex to real numbers. The result of simulating using float32 in the argument I mentioned is wildly different than using complex64; there is a lot more power traversing the region where there's complex permittivity, something which does not happen when using complex64. The latter appears to not have a lot of power going through it, at least until the very last time steps, where it shoots up.

This leads me to believe that permittivity is indeed being cast as a real number when using float32, and so the imaginary part is being discarded, thus losses in the material are not taken into account, making the simulation completely inaccurate. This should probably be mentioned as a warning when running a simulation, and/or in the documentation (haven't seen it in the latter but I haven't exhaustively read all of it). I am willing to open a pull request over this.

**Comment by ymahlau** [comment]
I just realized that I may have misunderstood your point. Are you saying that the simulation with complex numbers yields the correct result? This seems very surprising to me. Could you share the script you are working on?

**Comment by galcerte** [comment]
Sorry for the delay, but here's a simplified version of the script I'm using: [https://gist.github.com/galcerte/bae2a301f89f9dd6956578a74cf1c31b](https://gist.github.com/galcerte/bae2a301f89f9dd6956578a74cf1c31b) If you use either "complex64" or "float32" in the command line argument of the script (the command I used in the shell is `python sim_debug_permittivity.py --dtype="complex64" --gen_energy_video`), the output of the plots of the Poynting flux detectors should be markedly different between both dtypes.

I'm not quite sure that complex64 is correct; I am sure though that they are quite different. To me, it looks like the one using complex64 is correct, since the energy seems to "well up" inside the dielectric, and after some time it seems to saturate, when that happens, more power flows through. I am not used to working with lossy dielectrics though. This is not what happens in the float32 case at all, there the usual dielectric stuff seems to happen, which is just that the wave slows down and then gets out just fine. Maybe it's because I use `material=Material(permittivity=jnp.complex64(4.4 + 0.03 * 1j))` in `StaticMaterialObject`, that's one part of my code I didn't check too deeply, so I'm not sure if there's another way to define this.

**Comment by galcerte** [comment]
> Currently, we use a conductivity term in Maxwell's equations to be able to simulate complex permittivities or permeabilities. Therefore, it is not necessary to use a complex datatype in the simulation itself, because the real part (permittivity) and the complex part (conductivity) are modelled as separate arrays, each having data type float32 by default. You can set the conductivity in the material configuration of the respective objects.

Okay, so there's one thing I was missing. I am not specifying the conductivity for this material, but I am specifying its permittivity to be complex in the `permittivity` argument in `Material()`, and perhaps this is not supported.

**Comment by ymahlau** [comment]
I will add a check such that in the next release it will no longer be possible to set the dtype of the simulation to complex

**Comment by ymahlau** [comment]
I am not an expert with lossy materials either, but I am pretty sure you need to omit the $\varepsilon_0$. So the electric_conductivity parameter of the material should simply be $\sigma = \varepsilon''_r * \omega$. Admittedly i have not had the time to check our conduction model against an analytical solution. If you have any insights here please post them.

**Comment by galcerte** [comment]
I am consulting this with a colleague and he'll do the same with others. If I have anything worth sharing I'll make sure I do so.

**Comment by galcerte** [comment]
> You are right with the assumption that fdtdx was not build with these functionalities in mind, but I believe it is not a massive leap to implement most of them. To give a little context, fdtdx was originally created for designing 3d polymer structures, which is also my main research focus. For example, even a real permeability μ ≠ 1 was not considered in the first versions and only later implemented. I think it is just natural, that with having more people using fdtdx and more use-cases, fdtdx should grow to fit these demands. Also, conductivity was mainly implemented as a thought experiment to see if a gradient can be calculated through a lossy material by time reversibility (adding a gain term in the backward simulation).

Right, but I imagine it is not going to grow endlessly to fit every demand, either. Plasma for instance is one of the things I would guess FDTDX is never going to do, given that you need to plop an entirely different algorithm on top of FDTD which handles particle motion, and then you need to make the particles and the fields talk to each other. Though perhaps there is an application where $\rho_f \neq 0$ which doesn't involve plasma and the charges don't move...

> Regarding the extensions necessary to make these use cases work for fdtdx:
> 
>  Non-isotropic materials should be very easy to implement. This simply requires adding a new dimension of size 3 to the permittivity/permeability arrays representing the x/y/z components similar to the E/H arrays. Also, the API needs to be adjusted for specifying these materials. Isotropic materials can have a single value in this dimension and all calculations should still work due to the broadcasting rules in JAX.

That's great news. I saw the new issue, not sure if I have the time to contribute that at the moment, but I will try.

>  Truly complex permittivity should also be not difficult to implement, but due to my limited physics knowledge (I am coming from a CS background), I have not really a use case or idea how a simulation with that should look like. Do you have a specific example, maybe even with an analytical solution that could be used to check correctness? Regarding the implementation, I think it just boils down to using jnp.complex64 as a datatype and making sure that all the physics functions are able to handle complex values, since they were written without that thought in mind.

Well, if trying to simulate in the microwave, one use case is what I mentioned before: using FR-4. FR-4 is a fiberglass-reinforced epoxy laminate material used a lot in printed circuit boards (the green stuff you see). WiFi antennas emit in 2.4 GHz or 5 GHz, and those are etched onto FR-4. Another thing that would need this, this time closer to your original use case, is phase change materials (PCM) which do see use in nanophotonics (actually these would need to vary with time as well, since they change phase when heated, and thus permittivity...). However, as I mentioned in the previous post, you can pretty much just use conductivity, unless you have nonlinear, time-varying materials, weird nonlocal materials, or free charge density, then you really do need complex permittivity. It's just that the source of the rotational of the magnetic field has a different interpretation when using imaginary permittivity vs when using Ohm's law and conductivity, if considering Ampère-Maxwell in isolation. But if meeting the conditions I listed before, it's just that, an interpretation, the math remains the same.

By the way, when considering `jnp.complex64` as the data type for permittivity and such, would the fields themselves remain real, as in, using `jnp.float32`? Because if not, if they were represented using `jnp.complex64`, they'd take a lot more space and possibly take longer (already does in my testing actually).

>  A dispersion model is planned, but I think requires a bit more work (see [Dispersion Implementation #4](https://github.com/ymahlau/fdtdx/issues/4) for some ideas), but again I don't have any specific use case or example I could test this with. I think as a first step, a simple implementation of drude/lorentz would be a good starting point. As far as I have thought about it, the implementation basically also only requires adding the respective components as optional entries to the ArrayContainer class.

A very widely known example of dispersion would just be a prism, with incident light being broadband in the visible spectrum (so just white light). Not sure if anyone actually needs such a simulation, but it would be a good example to test against.

>  Non-zero free charge density, particle in cells algorithms or nonlinear materials are not planned (at least in the near future), because I think the use cases for these are rather specific. But feel free to correct me if you think that this is a feature we should consider.

No, it makes sense to me that these things are not planned, developer time and effort are not infinite after all. However, would you be willing to accept contributions to implement these features, as PRs? I might be interested in nonlinear materials. However, an actual complex permittivity implementation might be necessary if using these materials.

> The analytical solutions I was talking about was basically just how much energy of a uniform wave should be absorbed in a lossy material when travelling some specific distance. But when implementing the points above, I think we also need some more elaborate tests against analytical solutions.

Oh okay, that should be easy to make. ChatGPT o3 also gave me a few ideas as what to test against, and it was similarly simple. I imagine I would have to contribute such testing as pytest-compatible scripts that would go in the [tests/](https://github.com/ymahlau/fdtdx/blob/main/tests) folder, no?

**Comment by galcerte** [comment]
Okay, I'm now getting around to making the contributions I talked about. I have a few doubts: You said that I could write a tutorial and put a script in `examples`, so regarding this: do you mean that I make a script in that folder (duh) and that I write out information (like, say, what I wrote before in this issue) in the "Tutorials" section of the docs, within "Materials Guide"? That is, writing that out within `docs/tutorials/materials.md`? And should I make one branch for the example script and another for modifying docs?

**Comment by ymahlau** [comment]
I don't really have a strong opinion on this. Since I am in the process of reworking the documentation currently, maybe it would be a good idea to create something similar to [tidy3d's examples](https://www.flexcompute.com/tidy3d/examples/notebooks/AutoGrid/) there. I think using a jupyter notebook where you can write text as well as code execution is a really good way to structure a tutorial. Maybe you can add a new folder called ```notebooks``` and create a tutorial file called ```name_of_tutorial.ipynb``` there. Then you can have a single branch and a single PR for this one file. 

If you notice anything that could be improved on the fdtdx side in the process of making the tutorial feel free to open an issue on that. Or if you think you can fix it yourself create a separate PR for that. Suggestions for improvements are always welcome!

**Comment by galcerte** [comment]
Okay, I can do that, no problem. I'll let you know through other issues if I see anything that could be improved. I did feel like the documentation needs more things to be explained, since I needed to dig into the source code itself to get to understand how to use some classes, but since you're working on it, you must have everything that's needed into account.

**Comment by galcerte** [comment]
More things: in order to execute a notebook, it's necessary to install `pip` and `ipykernel` (not sure about `pip` as that might be a VS Code thing, but `ipykernel` is definitely a dependency). Should I put these packages in `pyproject.toml` in the value of `dependencies` or the value of `dev` under the `[dependency-groups]` table?

**Comment by ymahlau** [comment]
Not a dumb question, the simulation run will give you an ArrayContainer object, which has an attribute called detector_states. There you can find a dictionary mapping detector name -> detector state. Usually the detector state is also a dictionary containing the recorded values.

**Comment by galcerte** [comment]
Okay, thank you. I now have a bit of a first draft for the notebook: [lossy-material](https://github.com/galcerte/fdtdx/blob/lossy-material-example/notebooks/simulate_lossy_material.ipynb). I haven't uploaded any cell outputs, but I did run the notebook and for some reason I am getting quite a different value. The simulation yields 0.06 W through the detector, while the analytical expression yields 0.0000003 W. I need to check this in more detail to see if I made a mistake, which could be likely. I have a question or two: in the beginning, I put what I posted here in this issue, regarding conditions needed to identify conductivity with imaginary permittivity. I'm not sure if it does fit with the code in the notebook, since I am not really trying to model a lossy dielectric by varying `electric_conductivity` with frequency. So, I could use your opinion on whether it does fit or not.

What I *do* think should go here is the derivation of the analytical expression for the power flux, and I have not gotten around to writing that out. However, I shouldn't take long to do it.

Another thing I need your opinion on is how power flows through the detector. I have not fiddled with the PML thickness much, so I think that, because of this, I am observing a bit of a reflected wave on the bottom of the simulation volume where the wave impinges. I say this because power flux dips to negative values at its minimums, something which should definitely not happen if there is just an incident wave flowing in one direction.

**Comment by ymahlau** [comment]
I skimmed through the notebook, this looks already really good! Could you move the notebook into ```docs/notebooks```? I think that might actually be a better place for the notebook, sorry for the confusion.

Usually a pml of size 10 is sufficient for most applications, but if you want to be sure just set the size to 20. But then you should make sure that the simulation volume is still large enough and the source/detector are not within the pml. You can also make the whole simulation a bit more stable if you increase the ramp up time in the source (the time between starting to emit light and emitting light with full power). This can be done by specifying ```temporal_profile=fdtdx.SingleFrequencyProfile(num_startup_periods=5)``` (default is 2) in the source.

I think the conditions do fit into the notebooks. It is good to have the limitations of the software documented.

**Comment by ymahlau** [comment]
Oh that is probably a really important point! If the impedance is not set properly, then the source will not emit light properly, which usually manifests in power propagating in the opposite direction. That is probably the reason why you are measuring negative poynting flux. However, I am not sure if it even makes sense to place the source inside a lossy material. If I understand this [paper](https://arxiv.org/abs/1603.02720) correctly, then there are some weird effects happening when the incoming wave is in a lossy material. 

For the purpose of your tutorial, maybe it would be best to shrink the lossy material to (for example) 70% of the simulation volume and place the source in a non-lossy material above it. Then you can measure the poynting flux / energy at the beginning of the material very shortly after the interface and some distance later to compute the loss. 

Nonetheless, I will fix the impedance computation shortly to account for conductivity. Then you can also test if this makes a difference.

**Comment by ymahlau** [comment]
If you want to put the detector at the very edge of the interface, you can specify a grid margin of zero and place the detector at the top edge of the interface (relative position 1 or -1 in relative coordinates, depending on which side of the interface). This will put the detector in the first grid cells of the respective object. A grid margin of -1 would put the detector in the second grid cell behind the interface. 

That being said, I would generally advise to put at least a handful of grid cells between the interface and the detector to get an accurate measurement. Usually after 3-5 cells measurements are accurate.

**Comment by ymahlau** [comment]
This looks really good, you can open a PR if you want. Regarding the discrepancy, your notebook will be a good starting point to analyze where this discrepancy stems from and possibly eliminate this in future releases. 

Regarding you tanget: The ModeOverlapDetector measures the fourier transform of the fields. You can control if the whole detector volume is logged with the ```reduce_volume=False``` parameter (False is default). Maybe you set the parameter to True by accident? That would explain the logged shape.

**Comment by MarkCunningham0410** [comment]
Now that this is implemented, can hyperbolic anisotropic permittivities (i.e. eps_xx positive, eps_yy and eps_zz negative with real and imaginary components) be done across a range of frequencies?
